// nano-mcp/dist/nano-mcp/src/mcp-types.js
export class MCPServer {
    config;
    handlers;
    constructor(config) {
        this.config = config;
        this.handlers = new Map();
    }
    setRequestHandler(method, handler) {
        this.handlers.set(method, handler);
    }
    async start() {
        console.log(`Starting ${this.config.name} MCP server v${this.config.version}`);
        // Implementation would handle stdio communication
    }
    async handleMCPRequest(request) {
        try {
            const handler = this.handlers.get(request.method);
            if (!handler) {
                throw new Error(`Method ${request.method} not found`);
            }
            const result = await handler(request);
            return {
                jsonrpc: '2.0',
                result,
                id: request.id,
            };
        }
        catch (error) {
            return {
                jsonrpc: '2.0',
                error: {
                    code: -32603,
                    message: error instanceof Error ? error.message : String(error),
                },
                id: request.id,
            };
        }
    }
}
